### DISCUSSION TITLE
Roblox index nil to 'leaderstats'

### DISCUSSION THREAD

**Discussion by VolcanicGameStudios** [community/community#81376] [Programming Help] [answered]
### Body

So I have been trying to code a game and whenever I try to increase the value of one of the leaderstats it will say attemped to index nil 'leaderstats'.

Here is my code:

local Players = game:GetService("Players")
local dataStore = game:GetService("DataStoreService")
local database = dataStore:GetDataStore("data")
local sessionData = {}

local isAlive = true

local playerVar

function playerAdded(player)
	playerVar = player
	local leaderstats = Instance.new("Folder")
	leaderstats.Name = "leaderstats"
	
	local kills = Instance.new("IntValue")
	kills.Name = "Kills"
	kills.Parent = leaderstats
	
	local timeAlive = Instance.new("IntValue")
	timeAlive.Name = "Time Alive"
	timeAlive.Value = 0
	timeAlive.Parent = leaderstats
	
	
	local success = nil
	local playerData = nil
	local attempt = 1

	repeat
		success, playerData = pcall(function()
			return database:GetAsync(player.UserId)
		end)

		attempt += 1
		if not success then
			warn(playerData)
			task.wait(3)
		end
	until success or attempt == 5

	if success then
		print("Connected to database")
		if not playerData then
			print("Assigning defualt data")
			playerData = {
				["Kills"] = 0
			}
		end
		sessionData[player.UserId] = playerData
	else
		warn("Unable to get data for", player.UserId)
		player:Kick("Unable to load your data. Try again later")
	end

	kills.Value = sessionData[player.UserId].Kills
	
	
	leaderstats.Parent = player
	
	playerVar = player
	
end

Players.PlayerAdded:Connect(function(player)
	playerVar = player
	local leaderstats = Instance.new("Folder")
	leaderstats.Name = "leaderstats"

	local kills = Instance.new("IntValue")
	kills.Name = "Kills"
	kills.Parent = leaderstats

	local timeAlive = Instance.new("IntValue")
	timeAlive.Name = "Time Alive"
	timeAlive.Value = 0
	timeAlive.Parent = leaderstats

	local success = nil
	local playerData = nil
	local attempt = 1

	repeat
		success, playerData = pcall(function()
			return database:GetAsync(player.UserId)
		end)

		attempt += 1
		if not success then
			warn(playerData)
			task.wait(3)
		end
	until success or attempt == 5

	if success then
		print("Connected to database")
		if not playerData then
			print("Assigning defualt data")
			playerData = {
				["Kills"] = 0
			}
		end
		sessionData[player.UserId] = playerData
	else
		warn("Unable to get data for", player.UserId)
		player:Kick("Unable to load your data. Try again later")
	end

	kills.Value = sessionData[player.UserId].Kills

	leaderstats.Parent = player

	playerVar = player
end)

function gameloop(player)
	
	repeat
		task.wait(1)
		player.leaderstats["Time Alive"].Value += 1 --Problem here!
		
	until isAlive == false
end

gameloop(playerVar)

function playerLeft(player)
	
	if sessionData[player.UserId] then
		local success = nil
		local errorMsg = nil
		local attempt = 1
		
		repeat
			success, errorMsg = pcall(function()
				database:SetAsync(player.UserId, sessionData[player.UserId])
			end)
			
			attempt += 1
			if not success then
				warn(errorMsg)
				task.wait(3)
			end
		until success or attempt == 5
		
		if success then
			print("Data saved for", player.Name)
		else
			warn("Unable to save for", player.Name)
		end
		
	end
end

Players.PlayerRemoving:Connect(playerLeft)

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by VolcanicGameStudios** [comment, marked as answer]
nvm I fixed it

**Reply by ghostinhershell** [reply]
Thanks so much for following up and letting us know. And congrats on finding a solution. Happy coding! 🚀
